### PR TITLE
make: buildtest: fix parallel builds

### PR DESCRIPTION
**Pull request by kaspar030**
~~(CI test PR, please ignore)~~

I think I found the last parallel build errors.

This PR contains three commits:

cc96de02 makes "make buildtest" skip ccache on the first retry. Not sure it's needed.

91cec4c7 stops "make buildtest" from doing "rmdir $BINDIRBASE", which breaks parallel buildtest runs in the same source checkout

069bbcec make the micro-ecc package build within $BINDIR/pkg/micro-ecc. There was a race before which also broke parallel buildtests in the same source checkout

### CODE REVIEW

**Comment by kaspar030** [comment]
seems to be a heisenbug. :(

File: Makefile.base [hunk 1]
```diff
@@ -44,19 +44,22 @@ OBJ := $(OBJC) $(OBJCXX) $(ASMOBJ) $(ASSMOBJ)
...
 $(BINDIR)$(MODULE).a: $(OBJ) | ${DIRS:%=ALL--%}
+	@echo "ARCHIVE $(BINDIR)$(MODULE)" 2>&1
 	$(AD)$(AR) $(ARFLAGS) $@ $?
 
-
 CXXFLAGS = $(filter-out $(CXXUWFLAGS), $(CFLAGS)) $(CXXEXFLAGS)
 
 # compile and generate dependency info
 
 $(OBJC): $(BINDIR)$(MODULE)/%.o: %.c
```
Comment (Kijewski) [inline Makefile.base]: The module directory is missing? Try appending ` | $(BINDIR)$(MODULE)/` to the four `$(OBJ...)` goals.

Comment (kaspar030) [inline Makefile.base]: Shouldn't that be tackled by the lines
```OBJ := $(OBJC) $(OBJCXX) $(ASMOBJ) $(ASSMOBJ)```
and 
```$(BINDIR)$(MODULE).a $(OBJ): | $(BINDIR)$(MODULE)/```
?

Comment (Kijewski) [inline Makefile.base]: It should be, but maybe make evaluates something in the wrong order (?)

Comment (kaspar030) [inline Makefile.base]: I'd prefer that over a bug in tmpfs ;)
I just did > 5 successful builds with just an added "ls $BINDIR/$MODULE" after mkdir, all successful.
Now I removed that ls, to see if it fails again. If it does, I'll try adding the order deps as you suggested and see if it helps.

Comment (kaspar030) [inline Makefile.base]: The weird thing is, I have a build out with the "MKDIR xxx" line somewhere and further down a failed CC:
http://www.schleiser.de/~kaspar/fail.html

Comment (gebart) [inline Makefile.base]: Are any make clean targets called for riot ci? Are all groups being run in
the same tree with a shared output directory? Perhaps there is a race
cleaning up while another group is still building.

File: Makefile.buildtests [hunk 2]
```diff
@@ -38,12 +38,12 @@ buildtest:
 	@ \
 	BUILDTESTOK=true; \
 	APP_RETRY=0; \
-	rm -rf "$$BINDIRBASE"; \
 	for BOARD in $$($(MAKE) -s info-boards-supported); do \
```
Comment (kaspar030) [inline Makefile.buildtests]: That is probably the cause...

**Comment by gebart** [comment]
Good find!

File: Makefile.base (hunk 1 above)
Comment (kaspar030) [inline Makefile.base]: > Are any make clean targets called for riot ci? Are all groups being run
> in the same tree with a shared output directory? Perhaps there is a race
> cleaning up while another group is still building.

I think that's it, there was a "rm -rf $BINDIRBASE" at the beginning of
each "make buildtest" run.

**Comment by kaspar030** [comment]
> Good find!

You made me double-check. Thanks!

**Comment by kaspar030** [comment]
I'll run a couple of dozen test compiles ...

**Comment by gebart** [comment]
@kaspar030 could you try https://github.com/RIOT-OS/riotdocker/pull/17 at the same time?

File: Makefile.buildtests [hunk 3]
```diff
@@ -38,18 +38,18 @@ buildtest:
 	@ \
 	BUILDTESTOK=true; \
 	APP_RETRY=0; \
-	rm -rf "$$BINDIRBASE"; \
```
Comment (OlegHahm) [inline Makefile.buildtests]: make clean takes care of this, right?

Comment (kaspar030) [inline Makefile.buildtests]: Yep.
(```$BINDIRBASE``` would be "tests/unittests/bin2, while "make clean" only cleans "tests/unittests/bin/<board>", but for the build that makes no difference.)

File: Makefile.buildtests [hunk 4]
```diff
@@ -62,7 +62,7 @@ buildtest:
 	        WERROR=$${WERROR} \
 	        LTO=$${LTO} \
 	        TOOLCHAIN=$${TOOLCHAIN} \
-	        $(MAKE) -j$(NPROC) -B 2>&1) ; \
+	        $(MAKE) -j$(NPROC) clean all 2>&1) ; \
```
Comment (OlegHahm) [inline Makefile.buildtests]: would `-B clean all` do any harm here? Just to be on the safe side?

Comment (kaspar030) [inline Makefile.buildtests]: Does that make a difference? "clean" removes the whole build output directory, forcing everything to be rebuild. "-B" is totally redundant in that case. We've just added that some days ago...

Comment (OlegHahm) [inline Makefile.buildtests]: Hm, true...

**Comment by OlegHahm** [comment]
As far as I can tell, this shouldn't do any harm. ACK if Travis is fine, too (which seems to be the case).

**Comment by kaspar030** [comment]
Doesn't happen often that Mr.T and Murdock agree...

**Comment by OlegHahm** [comment]
Travis couldn't finish the last three tests from cortex_m3_1 group, because it got too big, too. But I think this should be fine enough for us.

**Comment by kaspar030** [comment]
Thanks for reviewing!

**Comment by OlegHahm** [comment]
Bullshit!

File: pkg/micro-ecc/Makefile [hunk 5]
```diff
@@ -16,28 +18,26 @@ MODULE:=$(shell basename $(CURDIR))
...
+	mkdir -p $(BINDIR)/pkg && \
 	git clone $(PKG_URL) $@ && \
 		cd $@ && git reset --hard $(PKG_VERSION)
 
 clean::
-	# Reset package to checkout state.
-	cd $(CURDIR)/$(PKG_NAME) || true && \
-		git clean -x -f && \
-		git reset --hard $(PKG_VERSION)
+	rm -Rf $(PKG_BUILDDIR)
```
Comment (gebart) [inline pkg/micro-ecc/Makefile]: why uppercase R here and lowercase r on distclean?

**Pull request merged**